### PR TITLE
[Ready] Batched cholesky decomposition

### PR DESCRIPTION
**Pull request by vishwakftw**
Implements batching for the Cholesky decomposition.

Performance could be improved with a dedicated batched `tril` and `triu` op, which is also impeding autograd operations.

Changes made:
- batching code
- tests in `test_torch.py`, `test_cuda.py` and `test_autograd.py`.
- doc string modification
- autograd modification
- removal of `_batch_potrf` in `MultivariateNormal`.

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -284,4 +301,67 @@ Tensor& potrs_out(Tensor& result, const Tensor& self, const Tensor& A, bool uppe
   return at::_th_potrs_single_out(result, self, A, upper);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ cholesky ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template<typename scalar_t>
+static void apply_cholesky(Tensor& self, bool upper, std::vector<int64_t>& infos) {
+#ifndef USE_LAPACK
+  AT_ERROR("cholesky: LAPACK library not found in compilation");
+#endif
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: nit: `#else` out the rest of the code?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -284,4 +301,67 @@ Tensor& potrs_out(Tensor& result, const Tensor& self, const Tensor& A, bool uppe
...
+  batchCheckErrors(infos, "cholesky");
+  return self_working_copy;
+}
+
+Tensor cholesky(const Tensor &self, bool upper) {
+  if (self.size(-1) == 0) {
+    return at::empty_like(self);
+  }
+  if (self.dim() == 2) {
+    return at::_th_potrf_single(self, upper);
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: So, at the end of the day, `_th_potrf_single` is calling the same LAPACK function, right? Can we just get rid of that codepath entirely and kill all of the TH bindings too? :)

Comment (ezyang) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: (Totes ok if that happens in another PR.)

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -284,4 +301,67 @@ Tensor& potrs_out(Tensor& result, const Tensor& self, const Tensor& A, bool uppe
...
+  }
+  squareCheckInputs(self);
+
+  // TODO: Once triu, tril is implemented for batched tensors, this
+  // can be simplified.
```
Comment (ezyang) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: nit: Maybe just write down explicitly what the simplification here is.

**Review by ezyang — APPROVED** [review]
Looks legit, thank you!

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (vishwakftw) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Yep, I think this should be possible. I'll get working on this soon.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (vishwakftw) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Soon, will do.

**Comment by vishwakftw** [comment]
@ezyang I have addressed comments and resolved merge conflicts too.

**Comment by vishwakftw** [comment]
Build failures are unrelated.

File: test/test_torch.py [hunk 4]
```diff
@@ -5334,6 +5334,29 @@ def test_cholesky(self):
...
+
+    @staticmethod
+    def _test_cholesky_batched(self, cast):
+        from common_utils import random_symmetric_pd_matrix
+
+        def cholesky_test_helper(n, batch_dims, cast, upper):
+            A = cast(random_symmetric_pd_matrix(n, *batch_dims))
+            cholesky_exp = torch.stack([m.cholesky(upper=upper) for m in A.reshape(-1, n, n)])
+            cholesky_exp = cholesky_exp.reshape_as(A)
+            print(torch.cholesky(A, upper=upper))
```
Comment (bddppq) [inline test/test_torch.py]: Should these debug lines be removed?

Comment (vishwakftw) [reply test/test_torch.py]: Oh, I missed them. I am sorry about this. I’ll send in a patch.

**Pull request closed without merging**